src/baseline.c

```diff
@@ -352,6 +353,137 @@ s8 lesq_solution_float(u8 num_dds_u8, const double *dd_obs, const double *N,
+  double new_obs[new_dds];
+  double new_N[new_dds];
+  double new_DE[new_dds * 3];
+
+  drop_i(dropped_dd, num_dds, 1, dd_obs, new_obs);
+  drop_i(dropped_dd, num_dds, 1, N, new_N);
+  drop_i(dropped_dd, num_dds, 3, DE, new_DE);
+
+  return lesq_solution_float(new_dds, new_obs, new_N, new_DE, b, resid);
+}
+
+/* Approximate chi square test
+ * Scales least squares residuals by estimated covariance, compares against threshold.
+ * Returns true if norm is below threshold.
+ *
+ * Returns residual value in `residual' if non-null.
+ */
+static bool chi_test(u8 num_dds, double *residuals, double *residual)
+{
+  double sigma = DEFAULT_PHASE_VAR_KF;
+  /* 5.5 seems adequate for float/fixed baseline residuals. */
+#define BASELINE_RESIDUAL_THRESHOLD 5.5
+  double norm = vector_norm(num_dds, residuals) / sqrt(sigma);
+  if (residual) {
+    *residual = norm;
+  }
+  return norm < BASELINE_RESIDUAL_THRESHOLD;
+}
+
+/* See lesq_solution_float for argument documentation
+ * Return values:
+ *    0: solution with all dd's ok
+ *    1: repaired solution, using one fewer observation
+ *       returns index of removed observation if removed_obs ptr is passed
+ *   -1: no reasonable solution possible
+ */
+/* TODO(dsk) update all call sites to use n_used as calculated here.
+ * TODO(dsk) add warn/info logging to call sites when repair occurs. */
+s8 lesq_solve_and_check(u8 num_dds_u8, const double *dd_obs,
+                        const double *N, const double *DE, double b[3],
+                        u8 *n_used,
+                        double *ret_residuals,
+                        u8 *removed_obs)
+{
+  integer num_dds = num_dds_u8;
+  double residuals[num_dds];
+
+  s8 okay = lesq_solution_float(num_dds_u8, dd_obs, N, DE, b, residuals);
+
+  if (okay == 0) {
+    double residual;
+    if (chi_test(num_dds, residuals, &residual)) {
+      /* Solution using all sats ok. */
+      if (ret_residuals) {
+        memcpy(ret_residuals, residuals, num_dds * sizeof(double));
+      }
+      if (n_used) {
+        *n_used = num_dds;
+      }
+      return 0;
+    } else {
+      if (num_dds < 5) {
+        /* We have just enough sats for a solution; can't search for solution
+         * after dropping one.
+         * 5 are needed because a 3 dimensional system is exactly constrained,
+         * so the bad measurement can't be detected.
+         */
+        if (n_used) {
+          *n_used = 0;
+        }
+        return -1;
+      } else {
+        u8 num_passing = 0;
+        u8 bad_sat = -1;
+        u8 new_dds = num_dds - 1;
+
+        for (u8 i = 0; i < num_dds; i++) {
+          lesq_without_i(i, num_dds, dd_obs, N, DE, b, residuals);
+          if (chi_test(new_dds, residuals, &residual)) {
+            num_passing++;
+            bad_sat = i;
+          }
+        }
+
+        if (num_passing == 1) {
+          /* bad_sat holds index of bad dd
+           * Return solution without bad_sat. */
+          /* Recalculate this solution. */
```

Comment:
We could store `b` and `residuals` for the `lesq_without_i` call that last passed the `chi_test` and then use them here to remove this expensive second call to `lesq_without_i`.